dev-clean/headers/llsc/filesys.hpp

```diff
@@ -3,152 +3,84 @@
-      ASSERT(c.size() == total_size, "Stat initialized with wrong sized vector.");
-      content = c;
-    }
-    Stat(const Stat& st): content(st.content) {}
-    immer::flex_vector<PtrVal> get_struct() const {
-      return content;
-    }
-    immer::flex_vector<PtrVal> read_field(st_field_t field) const {
-      unsigned pos, size;
-      std::tie(pos, size) = field_pos_size.at(field);
-      return content.drop(pos).take(size);
-    }
-    void write_field(st_field_t field, immer::flex_vector<PtrVal> c) {
-      unsigned pos, size;
-      std::tie(pos, size) = field_pos_size.at(field);
-      for (int i=0; i<size; i++) {
-        content = content.set(pos+i, c.at(i));
-      }
-    }
-};
+struct File: public Printable {
+    String name;
+    List<PtrVal> content;
+    List<PtrVal> stat;
+    // use transient? Need mutable map support in front end.
+    immer::map<String, Ptr<File>> children;
+    Ptr<File> parent;
 
-class File: public Printable {
-  private:
-    std::string name;
-    immer::flex_vector<PtrVal> content;
-    Stat stat;
-  public:
-    std::string toString() const override {
+    String toString() const override {
       std::ostringstream ss;
       ss << "File(name=" << name << ", content=[";
       for (auto ptrval: content) {
         ss << ptrval_to_string(ptrval) << ", ";
       }
+      ss << "], stat=[";
+      for (auto ptrval: stat) {
+        /* ss << ptrval_to_string(ptrval) << ", "; */
+      }
+      ss << "], nchildren=" << children.size();
+      ss << ", children=[";
+      for (auto &p: children) {
+        ss << p.first << ", ";
+      }
       ss << "])";
       return ss.str();
     }
-    File(std::string name): name(name) {}
-    File(std::string name, immer::flex_vector<PtrVal> content): name(name), content(content) {}
-    File(const File& f): name(f.name), content(f.content) {}
-
-    Stat& get_stat() {
-      return stat;
-    }
-
-    // if writing beyond the last byte, will simply append to the end without filling
-    void write_at_no_fill(immer::flex_vector<PtrVal> new_content, size_t pos) {
-      content = content.take(pos) + new_content + content.drop(pos + new_content.size());
-    }
-    void write_at(immer::flex_vector<PtrVal> new_content, size_t pos, PtrVal fill_val) {
-      int fill_size = pos - content.size();
-      if (fill_size > 0) {
-        // fill the new values to reflect the actual pos
-        content = content + immer::flex_vector(fill_size, fill_val);
-      }
-      write_at_no_fill(new_content, pos);
-    }
-    void append(immer::flex_vector<PtrVal> new_content) {
-      write_at_no_fill(new_content, content.size());
+    [[nodiscard]] inline static Ptr<File> create(String name) 
+    { 
+      return Ptr<File>(new File(name));
     }
-    void clear() {
-      content = immer::flex_vector<PtrVal>();
+    [[nodiscard]] inline static Ptr<File> create(String name, List<PtrVal> content)
+    {
+      return Ptr<File>(new File(name, content)); 
```

Comment:
it is usually better to use make_shared instead of calling its constructor explicitly

Reply:
Right. But in this case the constructors are made private, `make_shared` might not work. (See `struct Best` in [here](https://en.cppreference.com/w/cpp/memory/enable_shared_from_this))

Reply:
Not sure I understand why factory pattern (and making constructors private) would be necessary here -- those factory methods seem just direct wrappers of constructors, and there is only one File class without other hierarchy

Reply:
I guess it is not necessary, but it enforces the constraint that no stack objects can be created. Although if we never generate code like `File x1 = File()`, this should never happen, I still think that it is clearer to explicitly prohibit such usage.

Reply:
We can keep this conversation going and change later if another way turns out to be more appropriate. I will merge this PR for now since it is pretty large and has working implementation for a couple new syscalls.

Reply:
Ok, that sounds good!